### PR TITLE
Add interactive tiled region overlay

### LINKED ISSUE #62: Interactive visualization for tiled regions
**Section 1: Design details**

**Relevant background**

Due to the nature of the commercial instrument, the same problem of "deadlocking" (inability to adjust points) will appear for tiled ROIs on a sample as with TMA slides.

For tiling purposes, we need to provide a similar QC step in the TMA tiling process as with tiled region generation. This will involve creating an interactive visualization for users to verify where on the sample they have decided to tile and whether to delete any FOVs in the grid they have defined.

**Design overview**

The step-by-step process on the user's end associated with tasks that need to be addressed:

1. Co-registration: this already exists in `toffy`, it needs to happen for tiled regions over samples as well as TMAs.
2. Generate tiled regions: user will need to provide a tiled region input file defining the top-left corner of each FOV. We can then use the existing notebook to generate the FOVs for each tiled region. This will also handle randomization and Moly point insertion. The end result is a JSON that defines the initial run file.
    i. Erin has requested that we append the ROI identifier for each FOV name. So if there are 3 ROIs, named `'FOV1'`, `'FOV2'`, and `'FOV3'` in the spec file, her scheme will have them named `'FOV1_R1C1'`, `'FOV2_R1C1'`, `'FOV3_R1C1'`, etc. I think this is a nice addition as it will help identify which FOVs belong to which ROIs when visualizing them.
3. Interactive visualization: a lot of this code we can piggyback off of in `toffy`, especially the co-registration scheme. However, we'll need a dedicated function to draw the tiled regions since the following functionality will need to be added:
    i. Color-coding: each ROI should have a different color-coding scheme.
    ii. Ability to delete a FOV: done via a click. Once a FOV is clicked, the FOV will be deleted from the JSON defining the run file.
    iii. To prevent FOVs from overlapping on the image (happens when the ROIs are close to each other), the circles should be drawn small enough.
4. Saving the JSON to the run file: this can be done via a button in the interactive visualization

**Design list/flowchart**

tiled_region_flowchart drawio (1)

**Required inputs**

The following file inputs are needed:

- JSON spec file defining the top-left corners of each ROI to tile on the sample.
    - For each top-left FOV specified, the `centerPointMicrons` field and `fovSizeMicrons` field need to be defined. `centerPointMicrons` defines the centroid of the FOV and `fovSizeMicrons` defines the step size **along the row- and column-axes**.
    - Additionally, the `'name'` field needs to be defined for each top-left FOV. This will define the name of the ROI and the prefix of each FOV in that ROI during the tiling process
    - The user is responsible for ordering the FOVs in the correct ROI order
    - **Note that regardless of the name specified for these top-left FOVs in the spec file, we will treat these as `{ROI_Name}_R1C1`**.
- The path to the Moly point (if Moly point insertion is desired)
- Image defining the sample to be tiled
- Co-registration params file (generated by `update_coregistration_params.ipynb`)
- The name of the location to save the created run file info (in `.json` format)

Additionally, the tiling script already prompts the user to enter the number of FOVs along the row and column axes, as well as the randomization and Moly point insertion settings.

**Output files**

A run file defining the set of FOVs to extract.

**Section 2: Implementation details**

**Control flow**

_Generating the initial run JSON_

The tiling process as currently defined will remain the same, with the potential update of prefixing each fov name with `{ROI_Name}`.

_Interactive visualization updates_

When creating the interactive visualization, we can use the same co-registration as already exists for TMAs. The following additions will need to be made:

- Generate a unique color for each FOV depending on the ROI it belongs to. We can use a discrete colormap for this.
- Add interactivity to the dots indicating FOV labels. The TMA method of using `skimage` to draw the FOV dots directly on the slide will not work. We will need to overlay these dots directly on the image and use `mpl_connect` along with a custom circle drawing class to do this. Examples can be found here: [Matplotlib interactive examples](https://matplotlib.org/stable/users/explain/event_handling.html)
- We can use a similar default radius for each FOV dot on the image, it may need to be larger than for the TMA to account for the `{ROI_Name}` prefix on each FOV (ex. 10)

**Milestones and timeline**

Approximate timeline as follows:

1. Static overlay (2-3 days)
    i. Generating the TMA-like figure for the tiled regions
    ii. Implementing the circle drawing mechanism
    iii. Adding color-coding scheme per ROI
    iv. Potentially renaming the FOVs to include an `{roi_num}` prefix
2. Add dynamic features (1 week)
    i. Use `mpl_connect` to attach event listeners for `on_press` and `on_release`
    ii. Ensure `on_press` stores information about the specific FOV to delete, `on_release` removes the FOV dot from the visualization and from the run file JSON
    iii. Quality control (ex. no extra figures get drawn, each event is attached to the correct FOV dot, ensure the visualization works for different rectangle dimensions
3. Ensure save of run info to `.json` works properly (1 day)
    i. Piggyback off of the save annotation feature implemented for the TMA visualization

### PR DESCRIPTION
**Pull request by alex-l-kong**
**If you haven't already, please read through our [contributing guidelines](https://ark-analysis.readthedocs.io/en/latest/_rtd/contributing.html) before opening your PR**

**What is the purpose of this PR?**

Closes #62. The design doc is mostly intact, with a few divergences:

- FOVs will be drawn as a square instead of a circle
- There won't be any spacing between FOVs
- FOVs will only be outlined instead of fully shaded. On selection, the outline of the FOV will be darkened.

**How did you implement your changes**

See the design doc. The driver class is `FOVRectangle` which relies on `mpl_connect` functionality provided by `matplotlib`, allowing for interactivity.

**Remaining issues**

Mostly just aesthetics, the colormap may need to be changed, as well as the border width on selection.

Additionally, it's unfortunately not possible to test `ipywidgets` button functionality programmatically. See: https://github.com/jupyter-widgets/ipywidgets/issues/1896.

### CODE REVIEW

**Comment by alex-l-kong** [comment]
I think we have some open issues for this, but the main functionality should be good to go.

File: toffy/tiling_utils.py [hunk 1]
```diff
@@ -1271,11 +1273,311 @@ def generate_validation_annot(manual_to_auto_map, manual_auto_dist, check_dist=2
...
+        raise FileNotFoundError(
+            "Path %s to tiled_region_path does not exist, "
+            "please rename to a valid location" % os.path.split(tiled_region_path)[0]
+        )
+
+    # if there isn't a coreg_path defined, the user needs to run update_coregistration_params first
+    if not os.path.exists(os.path.join('..', 'toffy', 'coreg_params.json')):
+        raise FileNotFoundError(
+            "You haven't co-registered your slide yet. Please run "
+            "update_coregistraion_params.ipynb first."
```
Comment (ngreenwald) [inline toffy/tiling_utils.py:1445]: This is now the 1_set_up_toffy notebook. Can you also adjust the error message in the TMA script?

File: toffy/tiling_utils.py [hunk 2]
```diff
@@ -1271,11 +1273,311 @@ def generate_validation_annot(manual_to_auto_map, manual_auto_dist, check_dist=2
...
+
+        Args:
+            b (ipywidgets.widgets.widget_button.Button):
+                the button handler for `w_delete`, passed as a standard for `on_click` callback
+        """
+
+        delete_tiled_region_fovs(rectangles, tiled_region_fovs)
+
+    def save_mapping(b):
+        """Saves the mapping defined in `manual_to_auto_map`
```
Comment (ngreenwald) [inline toffy/tiling_utils.py:1481]: I think manual_to_auto_map was named

File: toffy/tiling_utils.py [hunk 3]
```diff
@@ -1271,11 +1273,311 @@ def generate_validation_annot(manual_to_auto_map, manual_auto_dist, check_dist=2
...
+
+        Args:
+            b (ipywidgets.widgets.widget_button.Button):
+                the button handler for `w_save`, passed as a standard for `on_click` callback
+        """
+
+        # needs to be in the output widget context to display status
+        with out:
+            # call the helper function to save manual_to_auto_map and notify user
```
Comment (ngreenwald) [inline toffy/tiling_utils.py:1490]: same here

File: toffy/tiling_utils_test.py [hunk 4]
```diff
@@ -421,16 +422,16 @@ def test_generate_tiled_region_fov_list(moly_path, moly_roi_setting,
...
+            (x, y) for y in list(reversed(np.arange(95, 105, 5))) for x in np.arange(0, 20, 5)
         ] + [
-            (x, y) for x in np.arange(50, 90, 10) for y in list(reversed(np.arange(140, 160, 10)))
+            (x, y) for y in list(reversed(np.arange(120, 160, 10))) for x in np.arange(50, 70, 10)
         ]
 
         # define the corresponding FOV names
         actual_fov_names = [
-            'TheFirstROI_R%dC%d' % (x, y) for y in np.arange(1, 3) for x in np.arange(1, 5)
+            'TheFirstROI_R%dC%d' % (x, y) for x in np.arange(1, 3) for y in np.arange(1, 5)
```
Comment (ngreenwald) [inline toffy/tiling_utils_test.py:432]: Does the logic for identifying region FOV names work if there are additional underscores in the user-defined name?

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looking good, couple small comments

File: toffy/tiling_utils.py [hunk 5]
```diff
@@ -287,7 +289,7 @@ def read_tiled_region_inputs(region_corners, region_params):
         region_params['region_start_row'].append(fov['centerPointMicrons']['y'])
         region_params['region_start_col'].append(fov['centerPointMicrons']['x'])
 
-        print("Using start coordinates of (%d, %d) in microns for region %s"
+        print("Using start coordinates of (%d, %d) in stage microns for region %s"
```
[the remark this answers is not in the document]
Comment (alex-l-kong) [reply toffy/tiling_utils.py:292]: I'll update this to just `microns`, probably leftover from when I was still getting used to the unit conversions between microns and stage coordinates and optical pixels.

File: toffy/tiling_utils_test.py (hunk 4 above)
Comment (alex-l-kong) [reply toffy/tiling_utils_test.py:432]: Yup, I just added that test case to `test_generate_tiled_region_fov_list`.

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks good. Can you find a time when Hadeesha has some spare time on the instrument and help him check out this branch and try it out?

**Comment by alex-l-kong** [comment]
@HPiyadasa how does this updated documentation look?

File: templates/2_create_tiled_mibi_run.ipynb [hunk 6]
```diff
@@ -13,7 +13,9 @@
    "cell_type": "markdown",
```
Comment (HPiyadasa) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>This is a bit confusing. </p><ol><li>Start with instructions on how to create the files and then where to place them. </li><li>The instructions say "the JSON file which defines each tiled region with their corresponding top-left corner FOVs". However its also followed by "In the 'fovs' list of that file, keep only FOVs you wish to define as a top-left corner for a distinct region" </li><li>The user should "name" the FOVs using Mibitracker before exporting the FOVs as it would make it easier than manually renaming the name key in the json file. </li><li>Why is it important to check the FOV parameters in the json file? </li></ol>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (HPiyadasa) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>This would be a bmp file if taken from D drive. Maybe just say "take optical image"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (HPiyadasa) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>png and bmp?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (HPiyadasa) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>Where can this be found? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (HPiyadasa) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>This should be at the top? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (HPiyadasa) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>Great!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (HPiyadasa) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>This link doesnt work</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: <p>Ah good catch, because we moved this over from <code>ark</code> our docs site no longer supports this. I'm not sure if we have plans to set up an equivalent docs site for this repo, so I'll just copy the documentation we have internally for this procedure here.</p>

Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: <p>Here's a new docstring that I think captures what you're saying:</p><p> </p><p>To identify each desired tiled region on the image, you must provide a <code>region_corners_path</code> that defines the top-left corner FOV for each of them.</p><p> </p><p>To generate these FOVs, use the Mibitracker to select the top-left FOV for each desired tiled region. Rename each FOV to the corresponding desired name of the tiled region it defines, then download the file (in JSON format). <code>region_corners_path</code> should be formatted just like a commercial instrument run file, with each top-left FOV contained in the <code>'fovs'</code> key. Place this file directly into the path specified in <code>json_tiling_dir</code>. Ensure this file is renamed <code>{tiled_region_prefix}_region_corners.json</code> (ex. if you set <code>tiled_region_prefix = test</code> in the notebook, the file should be renamed <code>test_tiled_region_corners.json</code>).</p><p> </p><p>For each top-left FOV, double check that the parameters <code>'name'</code>, <code>'centerPointMicrons'</code>, <code>'fovSizeMicrons'</code>, and the other parameters (ex. <code>'sectionId'</code>, <code> 'slideId'</code>, <code>'timingId'</code>) correspond with their values on the Mibitracker before using <code>region_corners_path</code> with the notebook. <code>'name'</code> defines the name of the tiled region; this must be correct because it is used to prefix the name of each FOV in the region to resolve ambiguity with FOV names in other tiled regions. <code>'centerPointMicrons'</code> and <code>'fovSizeMicrons'</code> are needed to position and size each FOV per tiled region; the same parameters defined in the top-left FOV will be used across all FOVs in the corresponding tiled region, so incorrect values will cause unwanted shifts and sizes for the entire region. The metadata parameters such as <code>'slideId'</code> are also copied from the top-left FOV to all other FOVs in its encompassing tiled region, so it's important that they're defined correctly for consistency purposes.</p>

Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: @HPiyadasa on second thought, I'm just gonna delete this. Users are likely not going to be needing Moly point separators anymore, and the description is straightforward enough to not need a separate link with additional info not contained in the existing documentation for this cell.

**Comment by alex-l-kong** [comment]
@ngreenwald tagging you as well for documentation and `matplotlib 3` updates to the visualization.

File: templates/2_create_tiled_mibi_run.ipynb (hunk 6 above)
Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>I would replace the description with something like this:</p><p> </p><p>"This notebook allows you to automatically set up large tiled runs of contiguous FOVs. You can specify the dimensions of the tiled image, for example 5x5, and select specific FOVs you wish to exclude from the final tile."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>What's the point of having the user create another directory? We should link to the readme explaining the directory structure. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>Bolding every word in every paragraph defeats the purpose of using bold. Please try and copy the style of the other notebooks. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>Mixing file names, instructions, and descriptions in this paragraph makes it extremely challenging to figure out what is going on. There's also extraneous information in here, like the description of the keys of the JSON file, which the user doesn't even know exists. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>The users aren't placing the variable <code>slide_path</code>, they're placing the actual slide image there.   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>We should delete this mention of moly_paths</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>Please shorten this to provide only the necessary information</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>There should be a GIF showing how to use this, to pair with extremely minimal instructions</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

**Review by ngreenwald — CHANGES_REQUESTED** [review]
A lot of the issues we identified in the other notebooks are present in this one. The idea behind going through all of those changes together was so that the same comments wouldn't need to keep being mentioned on each PR. 

Please try and incorporate the themes we talked about for clear, readable notebooks into this one, which right now has too much information and doesn't tell the user what they actually need to do at each step.

File: templates/2_create_tiled_mibi_run.ipynb (hunk 6 above)
Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: This was requested by @HPiyadasa, who wanted separate directories inside `tiled_regions_json` as a way to organize the the various region corners and slide files. This is similar to the problem we experienced during co-registration (addressed by #196), where users would put their copied notebook files into separate subdirectories inside of `templates` to eliminate clutter.

@ngreenwald I'll remove this for the time being. @HPiyadasa we can revisit this issue at a later time, but for the MIBI workshop it's might be easier to just use the main `tiled_regions_json` folder.

Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: @ngreenwald OK, I can remove the bolding. @HPiyadasa moving forward for the notebooks, it'll be easier to use normal markdown text with bolded words indicating important items, as opposed to different H1/H2 headings.

Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: <p>@ngreenwald as a follow up question, for the notebook documentation and slides, how necessary is it to go into detail about the method to select, name, and save the corner FOVs? Should we provide more handholding in this regard or assume a user of Mibitracker should know how to do this?</p><p> </p><p>If the latter, then this documentation can be simplified to a brief 2-3 sentence description of what <code>region_corners_path</code> is, how to name it, and where to place it.</p>

Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: <p>@ngreenwald oops, thanks for pointing that out.</p>

Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: <p>Related to issue #79.</p>

Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: <p>Should we also remove prompts for Moly points in <code>generate_tiled_region_fov_list</code>?</p>

Comment (ngreenwald) [reply templates/2_create_tiled_mibi_run.ipynb:13]: @HPiyadasa what was the issue that having a sub-folder is addressing?

Comment (ngreenwald) [reply templates/2_create_tiled_mibi_run.ipynb:13]: <p>Anyone using the notebook will know how to use the commercial instrument. So you won't need to tell them how to name an FOV, but you will need to tell what it needs to be named, etc. </p>

Comment (ngreenwald) [reply templates/2_create_tiled_mibi_run.ipynb:13]: <p>No, we'll keep that in case advanced users want to use it, but the default will be that they'll say no</p>

Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: Ok cool, @HPiyadasa we can revisit adding separate documentation for this at another time, you could open an issue for it.

Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: <p>Are we looking for a GIF per step (like the slides) or one GIF demonstrating a simplified walkthrough of the entire process?</p>

Comment (ngreenwald) [reply templates/2_create_tiled_mibi_run.ipynb:13]: <p>It looks to me like only bullet 1 needs a GIF. Up to you though, play around and see what is going to be easiest for a new user to understand who's never used the GUI before</p>

**Comment by alex-l-kong** [comment]
@ngreenwald hopefully documentation looks simpler and better, unfortunately playback speed for GIFs in Jupyter is really slow.

File: toffy/tiling_utils_test.py [hunk 7]
```diff
@@ -469,16 +471,16 @@ def test_generate_tiled_region_fov_list_base(moly_path, moly_roi_setting,
 
         # define the center points sorted
         actual_center_points_sorted = [
-            (x, y) for x in np.arange(0, 10, 5) for y in list(reversed(np.arange(85, 105, 5)))
+            (x, y) for y in list(reversed(np.arange(95, 105, 5))) for x in np.arange(0, 20, 5)
```
Comment (ngreenwald) [inline toffy/tiling_utils_test.py:474]: It seems like there's been a couple PRs where this has gotten changed. What's going on here?

**Review by ngreenwald — CHANGES_REQUESTED** [review]

File: templates/2_create_tiled_mibi_run.ipynb [hunk 8]
```diff
@@ -13,12 +13,14 @@
...
```
Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>I think this one got condensed a bit too much. When defining variables, I think the format from the previous version is good for users, since it visually separates each variable. </p><p> </p><p>Adding a visual layout for what the directory will look like is good idea, but if it's going to describe the directory structure, it should do just that, not also try and explain each file. It's also hard to tell that this is actual a visual layout of a directory structure, I would copy the format Sri used for the Mantis directory structure in the ark readme</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>It's not clear from the directions which cells the user needs to modify, and which can be run as-is. I would try and group together all of the lines that don't need modification into a smaller number of cells to reduce cognitive load on the user. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>This cell will automatically prompt you to... or something like that, so the user knows they'll need to run it</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>Does the user ever see the actual names for each FOV? If not, I don't think this information is that helpful</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>This cell can be combined with the one above</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>Content in the GIFs is good. Both of them are a bit too slow. </p><p> </p><p>Second GIF should have an example of selecting two FOVs, de-selecting one of them, then actually clicking the delete button</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:13]: <p>this can be one cell</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/86/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

File: toffy/tiling_utils_test.py (hunk 7 above)
Comment (alex-l-kong) [reply toffy/tiling_utils_test.py:474]: This change ensures that the center points we're testing against are sorted in MIBI coordinate order (specifically on the y-axis, which is decreasing from top to bottom).

File: templates/2_create_tiled_mibi_run.ipynb (hunk 8 above)
Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: They don't, I'll remove this. If the user really wanted to know the MIBI FOV naming standard they can check out the function or TMA notebook documentation.

Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: <p>Sounds good, I'm going to copy over the same format we have from <code>2_create_tma_mibi_run.ipynb</code> where we put the prefix up top, then have all the file definitions in one separate cell.</p><p> </p><p>Documentation for this section will also be organized like that section.</p>

Comment (alex-l-kong) [reply templates/2_create_tiled_mibi_run.ipynb:13]: <p>Now that I think about it, a visual directory structure image might be a bit much to add considering there are only a few files in one root directory, with no underlying subdirectory structure.</p><p> </p><p>The TMA notebook seems intuitive enough without one. Do you think this notebook can do without it as well, or should we keep it?</p>

Comment (ngreenwald) [reply templates/2_create_tiled_mibi_run.ipynb:13]: Up to you, whatever you think works best

**Review by ngreenwald — APPROVED** [review]
Looks great! Super understandable. Once Hadeesha is happy can merge it in.

**Comment by alex-l-kong** [comment]
@HPiyadasa did you have any comments on this one before merging in?

**Review by HPiyadasa — APPROVED** [review]
